oak-blob-composite/src/main/java/org/apache/jackrabbit/oak/blob/composite/DelegateHandler.java

```diff
@@ -0,0 +1,104 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.jackrabbit.oak.blob.composite;
+
+import org.apache.jackrabbit.core.data.DataIdentifier;
+import org.apache.jackrabbit.core.data.DataStore;
+import org.apache.jackrabbit.oak.spi.blob.DataStoreProvider;
+
+import java.util.Iterator;
+
+public interface DelegateHandler {
+    /**
+     * Add a {@link DelegateDataStore} to this handler.  This means the
+     * specified delegate will be considered a delegate by this delegate handler.
+     *
+     * @param dataStore A {@link DelegateDataStore} to be handled by this handler
+     */
+    void addDelegateDataStore(final DelegateDataStore dataStore);
+
+    /**
+     * Remove a {@link DelegateDataStore} from this handler.  This means the
+     * specified delegate will no longer be considered a delegate by this delegate
+     * handler.
+     *
+     * @param provider
+     * @return True if this action caused a {@link DataStoreProvider} to be removed,
+     * False if no matching {@link DataStoreProvider} was found.
+     */
+    boolean removeDelegateDataStore(final DataStoreProvider provider);
+
+    /**
+     * Determine whether this handler has at least one delegate.
+     *
+     * @return true if the handler has at least one delegate
+     */
+    boolean hasDelegate();
+
+    /**
+     * Get a data store iterator for all writable delegates managed by the
+     * {@link CompositeDataStore}.  They are returned in priority order based on the
+     * handler's prioritization strategy.
+     *
+     * @return Iterator to writable data stores
+     */
+    Iterator<DataStore> getWritableDelegatesIterator();
+
+    /**
+     * Get a data store iterator for all writable delegates managed by the
+     * {@link CompositeDataStore} that this handler believes contain a matching record
+     * for the provided identifier.  The iterator may contain no data stores if
+     * the handler believes the record doesn't exist in any writable delegate
+     * data stores.
+     *
+     * @param identifier A {@link DataIdentifier} for filtering the data stores
+     * @return Iterator to writable data stores
+     */
+    Iterator<DataStore> getWritableDelegatesIterator(final DataIdentifier identifier);
```

Comment:
I think what you are asserting here is based on a previous comment where you were asking whether there should be role-specific methods.  My understanding based on previous feedback was that roles should be free-form text.  In other words, I don't have a tightly defined "read-only" role.  Rather, I can name the role whatever I want, and then in the factory configuration for the delegate I tell it what role it has and specify any other configuration in there.

Reply:
Yes see my comment for that also. Essentially what I am asking is rather than have methods which are tied to specific role like `isReadOnly` or `getWritable` the methods should be more generic like: 
* `getRole(role)`
* `isRole(role)`
* `getDelegatesForRole(role)`
* `getDelegatesIteratorForRole(role)`

I understand that we don't have any use cases for others except read-only/write but if we are to support that in future then we'll have to change lot of things particularly the APIs as well.

Reply:
Aha Ok! I understand now. Ignore the above and my deleted comment